Integrates with Sleuth 2.2

adriancole opened the pull request:
This covers the data format from https://github.com/wavefrontHQ/wavefront-proxy/blob/3dd1fa11711a04de2d9d418e2269f0f9fb464f36/proxy/src/main/java/com/wavefront/agent/listeners/tracing/ZipkinPortUnificationHandler.java

This integrates with Sleuth 2.2, but doesn't use any new apis (would work even in Spring XML 2.5)

I've not been successful getting an account working with wavefront, it kicked out my vmware email and my attempt to register from my google account resulted in no email link (not trash etc)

Would appreciate someone with access verifying this as I noticed this project has no CI to actually integration test with the wavefront service, so hard to tell even before worked.

adriancole commented:
This project is working off spring-boot snapshot and I have no idea what the aspectj problem is, but someone who wants to try this out can use the following diff against https://github.com/openzipkin/sleuth-webmvc-example you'll notice this is basically drop-in and would still work even if the user are using OpenTracing instrumentation (bc sleuth provides a hook for that)

```diff
diff --git a/pom.xml b/pom.xml
index 71f99e2..5f8821e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -28,10 +28,16 @@
       <groupId>org.springframework.boot</groupId>
       <artifactId>spring-boot-starter-actuator</artifactId>
     </dependency>
-    <!-- Sends trace data to zipkin over http (defaults to http://localhost:9411/api/v2/spans) -->
     <dependency>
-      <groupId>org.springframework.cloud</groupId>
-      <artifactId>spring-cloud-starter-zipkin</artifactId>
+      <groupId>com.wavefront</groupId>
+      <artifactId>wavefront-spring-boot-starter</artifactId>
+      <version>2.0.0-SNAPSHOT</version>
+    </dependency>
+    <!-- TODO: figure out why everything dies on missing ReflectionWorld -->
+    <dependency>
+      <groupId>org.aspectj</groupId>
+      <artifactId>aspectjweaver</artifactId>
+      <version>1.9.5</version>
     </dependency>
   </dependencies>
 
diff --git a/src/main/resources/application.properties b/src/main/resources/application.properties
index bd50def..9b67cf4 100644
--- a/src/main/resources/application.properties
+++ b/src/main/resources/application.properties
@@ -5,3 +5,4 @@ spring.sleuth.traceId128=true
 spring.sleuth.sampler.probability=1.0
 # Adds trace and span IDs to logs (when a trace is in progress)
 logging.pattern.level=[%X{traceId}/%X{spanId}] %-5p [%t] %C{2} - %m%n
+spring.zipkin.sender.type=none
```

adriancole commented:
for those not familiar with sleuth, you can actually co-report to multiple different tracing backends `spring.zipkin.sender.type=none` intentionally turns off zipkin reporting, which declutters logs so you can focus only on the wavefront part.

adriancole commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingConfiguration.java:
```diff
@@ -1,36 +1,69 @@
+        wavefrontSender,
+        wavefrontConfig.source(),
+        createDefaultTags(applicationTags),
+        false /* TODO: spanLogsDisabled is an account setting. Maybe look it up? */
```
ps I'm mentioning this to be nice, not that this is a gap we should fill. This configuration as far as I can tell was not supported before. I just noticed in the proxy code, an account can have "span logs" disabled, and in that case it makes no sense to create overhead to send something that would be dropped on the other side. IOTW this is an advisory comment not a "required before merge" imho.

panghy replied:
the SDK currently will drop all span logs after the server refuses them permanently. the intention is to have this feature enabled on all clusters though

adriancole replied:
great. we can take this out of scope of this abstraction then. Thanks!

adriancole commented on wavefront-spring-boot/src/test/java/com/wavefront/spring/autoconfigure/IntegratedTracingTests.java:
```diff
@@ -0,0 +1,167 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(
```
note: this is integration test in so far as usually meant in projects like sleuth, but not actually hitting the service.

I would highly recommend some sort of CI integration with an actual account that reads back data as otherwise anything here (and elsewhere in the project) could break and no-one would know due to the async nature of telemetry reporting.

This doesn't have to be fancy either, for example, the spring-cloud-gcp team made a little shell script in a similar project, it just does a couple commands, but it helps verify sanity without clicking to do the same

https://github.com/openzipkin/zipkin-gcp/blob/master/travis/publish.sh#L107-L160

adriancole commented:
It seems I was confused by the micrometer setup, which isn't actually using a bean for the metrics sender. hence tracing can't share it.

Ex. I debugged and found this implicit connection being made inside the bean method.
```java
	@Bean
	@ConditionalOnMissingBean
	public WavefrontMeterRegistry wavefrontMeterRegistry(WavefrontConfig wavefrontConfig, Clock clock) {
		return WavefrontMeterRegistry.builder(wavefrontConfig).clock(clock).httpClient(
				new HttpUrlConnectionSender(this.properties.getConnectTimeout(), this.properties.getReadTimeout()))
				.build();
	}
```

adriancole commented:
It seems spring-boot-actuator 2.2.5 brings in a different auto-configuration for micrometer, which doesn't expose the sender bean, while the snapshot version does expose it.

I will change the example app to manually override the spring boot version in hopes it catches the right actuator.. and see if it works via debugger.

@snicoll this is the sort of reason I was not a big fan of bringing sleuth which is based on 2.2 as a dep when the project here only tests against boot 2.3

adriancole commented:
the sender thing is ok only using latest spring boot snapshot.. progressing further now.

snicoll commented:
> It seems spring-boot-actuator 2.2.5 brings in a different auto-configuration for micrometer, which doesn't expose the sender bean, while the snapshot version does expose it.

@adriancole you need [Spring Boot `2.3.0.BUILD-SNAPSHOT`](https://github.com/wavefrontHQ/wavefront-spring-boot#prerequisites). I've tried petclinic against your PR and I don't see any traces in Wavefront FTR.

adriancole commented:
really wish I had a login.. it is annoying to try to debug only with a debugger

snicoll commented:
I can give you a login @adriancole please reach out in DM

adriancole commented:
https://github.com/openzipkin/sleuth-webmvc-example/compare/wavefront try this also

I think the last glitch after snapshot was sleuth set sampling only when zipkin was in the classpath. it might work now. I'll follow-up on that glitch later

adriancole commented:
handing this over to @snicoll as things communicate now (using the sleuth-webmvc-example above and this branch)

![Screenshot 2020-04-18 at 10 24 04 PM](https://user-images.githubusercontent.com/64215/79640283-500f3080-81c3-11ea-9ee5-290d5022004b.png)

snicoll commented:
Thanks a lot @adriancole. I've got tracing as well with petclinic.

Integrating with Spring Cloud revealed an issue with the context boostrap that I need to workaround, see https://github.com/spring-cloud/spring-cloud-commons/issues/737

Adrian and I also discuss the `@Import` he added on  `SamplerAutoConfiguration`. I need to figure out why that is and how we can avoid that. I'll submit PRs to that affect.

adriancole commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontSpanHandler.java:
```diff
@@ -0,0 +1,169 @@
+      tags.add(Pair.of("ipv4", span.localIp())); // NOTE: this could be IPv6!!
+    }
+
+    // Note: this will block the caller. For example, it will delay production requests. Consider
```
probably we can remove this comment as I don't think it likely anyone will have an implementation that is blocking.

snicoll commented:
I've added a fix for the context bootstrap problem in #24

adriancole commented:
incidentally, I looked and the the handler part of this code would work with Sleuth all the way back to 2.0, as well even [Spring 2.5 XML](https://github.com/openzipkin/brave-webmvc-example/tree/master/webmvc25).

As folks are in a focused effort towards the upcoming RC, there's probably not enough energy or immediate value to extract this, yet. Just mentioning that if something comes up in the future, it can come in handy. In practice, we routinely get questions of sites mixed between boot and spring XML, and that's one large reason Sleuth uses Brave.

snicoll commented:
For other reasons we need Spring  Boot 2.3 and Micrometer 1.5 so I think we are fine.

Here is how I would like to get this merged, which I believe is a good summary of our meeting yesterday:

* If Sleuth is enabled for the app, use that and export traces based on the proposal in this PR. Sleuth is what the Spring ecosystem will expect so we need to make sure it is a first class citizen
* This project has historically OT support and customers are using that. If Sleuth is not configured, fallback to what we currently have on `master`.

@panghy can I get a review of the wavefront team on this please? If you need my help to setup a sample, please let me know. I am looking specifically at the `WavefrontSpanHandler` and what is actually exported to Wavefront. I've noticed the dashboard is less rich when the app runs with this PR and Sleuth so I'd love us to collaborate to provide the same information, if possible.

Once I have a review, I will work on the auto-configuration part and the tests, the same way I did for this project initially. I've fixed the Spring Cloud limitation in #24 and I still need to figure out why we have to import an auto-configuration but hopefully we'll find a fix for that too.

Can you give me an ETA for the review? Thanks a lot!

adriancole commented:
ps sleuth has also had OT users for years https://cloud.spring.io/spring-cloud-sleuth/reference/html/#opentracing

snicoll commented:
That's good to know! Something to keep in mind.

@adriancole I went to the bottom of why you need to add that import on `SamplerAutoConfiguration` but I am not sure what to recommend at this point. You have mentioned that to me previously but it didn't click.

`SamplerAutoConfiguration` and `TraceAutoConfiguration` are both competing for providing a `brave.sampler.Sampler`. `TraceAutoConfiguration` provides `Sampler.NEVER_SAMPLE` [by default](https://github.com/spring-cloud/spring-cloud-sleuth/blob/89ac3ed4aa5e4de2654be6c33f152e4dafd94fba/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java#L138-L142). `SamplerAutoConfiguration`, however, provides an actual `Sampler`. If the latter isn't contributed to the context, no trace gets exposed which is consistent with the discussion above.

Unfortunately, there is no ordering between those two auto-configurations and I am not even sure what the intention is there really. `TraceAutoConfiguration` runs first, creates  the no-op `Sampler`. `SamplerAutoConfiguration` runs next and backs off because a `Sampler` bean is already defined. 

The import you added in this PR overrides the order. Do you know why Sleuth has two sources for such a bean? How is `SamplerAutoConfiguration` supposed to work since `TraceAutoConfiguration` provides a no-op bean?

One way to mitigate the issue would be to create a `Sampler` bean ourselves here and make sure we run fist but then we lose the logic about the `@RefreshScope`.

adriancole commented:
@snicoll terse because on a call, not because of you. don't read into the way it exists. SamplerAutoConfiguration should control the sampler not TraceAutoConfiguration. If there's a nice way to decouple responsibility we should change to that in sleuth and then rely on tests that exists to prove it. I'm not personally good enough at spring boot to know  if that means AutoConfigureAfter or something else. what you think is best.. we should try, but meanwhile probably pragmatic to import here I think with a comment on the PR to sleuth to fix it. Wanna raise a PR to unspaghetti this (in sleuth) otherwise tell me best practice and I'll do it.

panghy commented:
So if I read you correctly @snicoll, we would be switching out sleuth and OT by default and if the user plops in the dependency, we'll detect it on the classpath and configure right?

I can take a look at this.

panghy commented:
So what I think is missing is the metricization that happens in the OpenTracing SDK that we have and without that, the landing page that we plop people on (the service page) will not show the right things (you can only see traces via the "All Traces" link).

@sushantdewan123 is there some common code from the OT tracer that we can use to do that without reinventing the wheel here?

snicoll commented:
@panghy I have already something in that direction locally. I am waiting for a review of this PR to massage things a bit so that we can do the right thing based on the classpath. I didn't review the bits Adrian wrote with regard to tracing itself (i.e. `WavefrontSpanHandler`) so a review of that would be great. Once I have a green light of the team, I can fine tune and merge this one.

IMO the priority is to make sure that both solutions provides a similar user experience. If you can look into that, that would be great!

adriancole commented:
> So what I think is missing is the metricization that happens in the OpenTracing SDK that we have and without that, the landing page that we plop people on (the service page) will not show the right things (you can only see traces via the "All Traces" link).

hmm interesting.. are you doing metrics on spans? does that imply you are doing 100% sampling? If so, it simplifies another discussion. Regardless brave has a concept of "local sampled" which means that you can do 100% aggregation of operations even if they are not remotely sampled.

Here's a simplified example of micrometer which I presume the SDK code, once understood, could be ported to https://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/features/handler/SpanMetricsCustomizer.java

snicoll commented:
> SamplerAutoConfiguration should control the sampler not TraceAutoConfiguration. If there's a nice way to decouple responsibility we should change to that in sleuth and then rely on tests that exists to prove it.

Thanks Adrian, I've moved that conversation to https://github.com/spring-cloud/spring-cloud-sleuth/issues/1618

adriancole commented:
I fixed the config bug where service name wasn't coming through properly. I also added a note as usually spring users like to have least config possible and explicitly setting `wavefront.application.name=${spring.application.name}` will be a bore to people not already using wavefront.

Further, I know there is some doubt that persists about sleuth and opentracing. I rebased my example branch over the existing sleuth opentracing example to prove it works.

https://github.com/openzipkin/sleuth-webmvc-example/compare/wavefront

![Screenshot 2020-04-23 at 12 15 06 PM](https://user-images.githubusercontent.com/64215/80058590-450a2680-855c-11ea-9a99-3b157de2f7b7.png)


Note: we really really should not advise or otherwise pull in configuration from opentracing starters as they are often copies of sleuth in the first place, and as well having duplicate instrumentation will be messy. In other words let's please not begin a problem we don't currently have in sleuth. Right now, users only use OpenTracing for things sleuth is missing, notably custom tracing code.

adriancole commented:
At first I thought the metrics you were referring to were the health metrics. I integrated them anyway, but I believe it must be the RED metrics. I think I can figure it out from the other code

adriancole commented:
and yeah as far as I can tell the RED metrics in wavefront are not on sampled data, which makes sense, but I wanted to clarify.

adriancole commented:
I think I figured out the jist of what's going on with RED metrics by reverse engineering the opentracing code.

It is probably nonsense to layer anything else onto this PR until it is likely to be merged. FWIW here's a rough cut https://github.com/adriancole/wavefront-spring-boot/compare/integrate-sleuth...adriancole:integrate-sleuth-red

panghy commented:
I believe what might still be missing is the component heart-beat metrics that are sent as long as it is initialized. @sushantdewan123 can add to this. We don't have a component for sleuth today (or would it be brave?) in the backend but we can certainly add that (essentially recognizing the component metric and showing the service).

Then again, maybe we should just call it the spring boot component? This has implications in the future since Jaegar and Zipkin are just jaegar and zipkin component to us because we don't have a way to tell what's generating them but for Sleuth, should we capture the actual component so that we can customize the landing page as well?

adriancole commented:
IMHO both the heartbeat and RED metrics stuff are more metrics/micrometer than sleuth brave zipkin or jaeger (or otherwise).

Ex Maybe heartbeat and RED could be callback of of two timestamps and a micrometer tag set?

The interface could say "give me these things for every span in the firehose, regardless of the api implementing the firehose.

Put more specifically, https://github.com/adriancole/wavefront-spring-boot/compare/integrate-sleuth...adriancole:integrate-sleuth-red is mostly just micrometer code... that could be a component indeed.

panghy commented:
@snicoll thoughts? this is definitely a tracing/metrics interaction whereby we need:

1) component heartbeats (regardless of whether there is traffic) and
2) metricization of spans

Bonus points for 1) to send heartbeats on start-up (once and resume 60s) because then we can immediately draw the proper UI elements. Perhaps that's a question for micrometer (like force a flush after start-up).

Perhaps we can add a WavefrontSender shim that does 2)? thoughts @sushantdewan123 

For 1), we'll need to have an inventory of components somehow.

sushantdewan123 commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontSpanHandler.java:
```diff
@@ -0,0 +1,224 @@
+ */
+final class WavefrontSpanHandler extends FinishedSpanHandler implements Runnable, Closeable {
+  private static final Log LOG = LogFactory.getLog(WavefrontSpanHandler.class);
+  // https://github.com/wavefrontHQ/wavefront-proxy/blob/3dd1fa11711a04de2d9d418e2269f0f9fb464f36/proxy/src/main/java/com/wavefront/agent/listeners/tracing/ZipkinPortUnificationHandler.java#L113-L114
```
minor nit:
this should be #L114-L114

sushantdewan123 commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontSpanHandler.java:
```diff
@@ -0,0 +1,224 @@
+    meterRegistry.gauge("reporter.queue.size", spanBuffer, sb -> (double) sb.size());
+    meterRegistry.gauge("reporter.queue.remaining_capacity", spanBuffer,
+        sb -> (double) sb.remainingCapacity());
+    this.spansReceived = meterRegistry.counter("reporter.spans.received");
```
we usually have prefix ~sdk
so final name that we want to report to wavefront would be with the following prefix
`~sdk.spring-boot.*`
so you can rename all the reported metrics with that prefix
for instance 
"~sdk.spring-boot.reporter.spans.received"

snicoll replied:
This feels odd to me to put "spring boot" in the metrics. Can you share why you need that information?

sushantdewan123 commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontSpanHandler.java:
```diff
@@ -0,0 +1,224 @@
+
+    // Start and duration become 0L if unset. Any positive duration rounds up to 1 millis.
+    long start = span.startTimestamp() / 1000L, finish = span.finishTimestamp() / 1000L;
+    long duration = start != 0 && finish != 0L ? Math.max(finish - start, 1L) : 0L;
```
rename to durationMillis

sushantdewan123 commented on wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontSpanHandler.java:
```diff
@@ -0,0 +1,224 @@
+    if (name == null) name = DEFAULT_SPAN_NAME;
+
+    // Start and duration become 0L if unset. Any positive duration rounds up to 1 millis.
+    long start = span.startTimestamp() / 1000L, finish = span.finishTimestamp() / 1000L;
```
rename to startMillis and finishMillis

adriancole commented:
ps @snicoll (who is in a better timezone anyway) offered to polish this up and do follow-ups when he's free. I really didn't initially intend to spend a dozens of hours on this, though that's past :P

snicoll commented:
Thanks a lot for that @adriancole. I've polished as discussed so that Spring Cloud Sleuth is configured and a fallback configuration with OT is applied if it isn't around. 

@adriancole I didn't do anything with the RED metrics [you've started working on](https://github.com/adriancole/wavefront-spring-boot/commit/71c4b8590bfcf076b987269674f4245e604917b7). Would you be so kind to rebase that work from `master` and submit another PR so that we can discuss that part?

@sushantdewan123 I've applied your review, thank you very much. I didn't change the prefix of the metrics as I didn't understand why they had to have `spring-boot` in their names.

adriancole commented:
@snicoll Thanks for taking this on

here's the other thing you asked about https://github.com/wavefrontHQ/wavefront-spring-boot/pull/35

To finish this after incorporating feedback might be similar to the integration test I made here, except not using a WF span service fake, rather one for the existing heartbeat and red metrics api.
